torch/utils/data/common.py

```diff
@@ -0,0 +1,162 @@
+        extract_tars : bool = True,
+        extract_zips : bool = True,
+        recursive : bool = True,
+        mask_tars : Union[str, List[str]] = '',
+        mask_zips : Union[str, List[str]] = ''):
+
+    # test whether binary_stream is seekable (eg. PIPE stream from webdata is not seekable)
+    seekable = hasattr(binary_stream, "seekable") and binary_stream.seekable()
+
+    # Note: We have no way to verify whether a non-seekable stream (eg. PIPE stream) is tar/zip without
+    #       changing stream handle position, however, there is no way to move such stream's handle back.
+    #       For non-seekable stream, the entire stream will be discard if it can not be extracted
+
+    # extract tar file
+    if extract_tars and match_masks(pathname, mask_tars):
+        try:
+            with tarfile.open(fileobj=binary_stream) as tar:
+                for tarinfo in tar:
+                    if not tarinfo.isfile():
+                        continue
+
+                    extract_fobj = tar.extractfile(tarinfo)
+                    if extract_fobj is None:
+                        warnings.warn("failed to extract tar file {}".format(tarinfo.name))
+                        continue
+
+                    inner_pathname = os.path.normpath(os.path.join(pathname, tarinfo.name))
+                    if recursive:
+                        yield from extract_files_from_single_pathname_binary(
+                            inner_pathname, extract_fobj, extract_tars, extract_zips, recursive, mask_tars, mask_zips)
+                    else:
+                        yield (inner_pathname, extract_fobj)
+            return
+        except tarfile.TarError:
+            if not seekable:
+                warnings.warn("Unable to reset the non-tarfile stream {}, skip!".format(pathname))
+                return
+            binary_stream.seek(0)
+
+    # extract zip file
+    if extract_zips and match_masks(pathname, mask_zips):
+        try:
+            with zipfile.ZipFile(binary_stream) as zips:
+                for zipinfo in zips.infolist():
+                    # major version should always be 3 here.
+                    if (sys.version_info[1] < 6 and zipinfo.filename.endswith('/')) or zipinfo.is_dir():
+                        continue
+
+                    inner_pathname = os.path.normpath(os.path.join(pathname, zipinfo.filename))
+                    if recursive:
+                        yield from extract_files_from_single_pathname_binary(
+                            inner_pathname, zips.open(zipinfo), extract_tars, extract_zips, recursive, mask_tars, mask_zips)
+                    else:
+                        yield (inner_pathname, zips.open(zipinfo))
+            return
+        except zipfile.BadZipFile:
+            if not seekable:
+                warnings.warn("Unable to reset the non-zip stream {}, skip!".format(pathname))
+                return
+            binary_stream.seek(0)
+
+    yield (pathname, binary_stream)
```

Comment:
Wait, here you will yield input stream AFTER yielding all unpacked files, please remove this part.

Reply:
I think this is as designed. We will only touch this piece when input steam is not a supported zip format. (not tar/zip), and we return the original binary stream in this case.